clients/drcachesim/tests/scheduler_unit_tests.cpp

```diff
@@ -987,6 +987,74 @@ test_synthetic_with_priorities()
+    static constexpr int NUM_INPUTS_PER_WORKLOAD = 3;
+    static constexpr int NUM_OUTPUTS = 5;
+    static constexpr int NUM_INSTRS = 9;
+    static constexpr memref_tid_t TID_BASE = 100;
+    std::vector<scheduler_t::input_workload_t> sched_inputs;
+    auto get_tid = [&](int workload_idx, int input_idx) {
+        return TID_BASE + workload_idx * NUM_INPUTS_PER_WORKLOAD + input_idx;
+    };
+    for (int workload_idx = 0; workload_idx < NUM_WORKLOADS; workload_idx++) {
+        std::vector<scheduler_t::input_reader_t> readers;
+        for (int input_idx = 0; input_idx < NUM_INPUTS_PER_WORKLOAD; input_idx++) {
+            memref_tid_t tid = get_tid(workload_idx, input_idx);
+            std::vector<trace_entry_t> inputs;
+            inputs.push_back(make_thread(tid));
+            inputs.push_back(make_pid(1));
+            for (int instr_idx = 0; instr_idx < NUM_INSTRS; instr_idx++) {
+                // Include timestamps but keep each workload with the same time to
+                // avoid complicating the test.
+                if (instr_idx % 2 == 0) {
+                    inputs.push_back(make_timestamp(10 * (instr_idx + 1)));
+                }
+                inputs.push_back(make_instr(42 + instr_idx * 4));
+            }
+            inputs.push_back(make_exit(tid));
+            readers.emplace_back(
+                std::unique_ptr<mock_reader_t>(new mock_reader_t(inputs)),
+                std::unique_ptr<mock_reader_t>(new mock_reader_t()), tid);
+        }
+        sched_inputs.emplace_back(std::move(readers));
+        // We do a static partitionining of the cores for our workloads with one
+        // of them overlapping the others.
+        std::set<scheduler_t::output_ordinal_t> cores;
+        switch (workload_idx) {
+        case 0: cores.insert({ 2, 4 }); break;
+        case 1: cores.insert({ 0, 1 }); break;
+        case 2: cores.insert({ 1, 2, 3 }); break;
+        default: assert(false);
+        }
+        sched_inputs.back().thread_modifiers.emplace_back(cores);
+    }
+
+    scheduler_t::scheduler_options_t sched_ops(scheduler_t::MAP_TO_ANY_OUTPUT,
+                                               scheduler_t::DEPENDENCY_TIMESTAMPS,
+                                               scheduler_t::SCHEDULER_DEFAULTS,
+                                               /*verbosity=*/3);
+    sched_ops.quantum_duration = 3;
+    scheduler_t scheduler;
+    if (scheduler.init(sched_inputs, NUM_OUTPUTS, sched_ops) !=
+        scheduler_t::STATUS_SUCCESS)
+        assert(false);
+    std::vector<std::string> sched_as_string =
+        run_lockstep_simulation(scheduler, NUM_OUTPUTS, TID_BASE);
+    for (int i = 0; i < NUM_OUTPUTS; i++) {
+        std::cerr << "cpu #" << i << " schedule: " << sched_as_string[i] << "\n";
+    }
+    // We have {A,B,C} on {2,4}, {D,E,F} on {0,1}, and {G,H,I} on {1,2,3}:
+    assert(sched_as_string[0] == "DDDFFFDDDFFFDDDFFF");
+    assert(sched_as_string[1] == "EEEHHHEEEIIIEEE");
```

Comment:
The first input scheduled on core 1: I see it is E, why not G? G is eligible and also has a lower workload-local timestamp than E. Then on core 3, you'll have H.

Reply:
These actually all have the same workload-local timestamps (this doesn't use the weighted timestamps that the other test has) which makes it easier to reason about the results here.

Reply:
But if it did use weighted timestamps would we see a different schedule (G instead of E)?

Reply:
The timestamps from the other test are inverted to G would have later timestamps than E being 1st vs 2nd in the triplet.  I tried it and actually it found a bug in pop_from_ready_queue() where if the q is empty it thinks there's nothing, skipping the last one: fixed now.  We then end up with:
```
cpu #0 schedule: FFFFFFFFFEEEEEEEEE
cpu #1 schedule: IIIIIIIIIDDDDDDDDD
cpu #2 schedule: CCCCCCCCCAAAAAAAAA
cpu #3 schedule: HHHHHHHHHGGGGGGGGG
cpu #4 schedule: BBBBBBBBB
```
It is less interesting as we get no interleaving.

Reply:
I'm adding this as a separate test to include that final entry case.